MSC4155: Invite filtering (#4155)
matrix-org/matrix-doc: proposals/4155-invite-filtering.md

Reviewer: Implementation requirements:
* [ ] Client supporting the feature, and using it

Author: Since I believe all the gematik implementations will be closed source, I'll reference https://github.com/matrix-org/matrix-spec-proposals/pull/3860#discussion_r1170031289 as an example for how cases like this were handled in the past. Thanks @clokep for digging it up.

Reviewer 2: https://github.com/element-hq/synapse/pull/18288 is a serverside implementation

Reviewer 2: https://github.com/element-hq/element-web/pull/29603 exposes the serverside settings in the client, but does no filtering of itself. @Johennes does this evoke any worries from you if the invite filtering is done server-side?

Reviewer 2: Yep, I included that in the Synapse side. Albeit, it's just a versions flag to say that it's supported. It might be better as a capability tbh.

Author: Yeah, I think a capability would be better eventually. But not a blocker for the PoC implementation. Thanks for pulling this off. I guess I will have to switch back to EW nightly now. 😅

Reviewer 2: Well, I still need to get all this stuff approved but hopefully yes :). EWeb has lots of nice safety features landing :)

Reviewer 3: Similar implementations:

- Implementing the gematik spec: https://github.com/famedly/synapse-invite-checker/blob/5608f40ffeaef1320edef8554cb432fa2c881fdd/synapse_invite_checker/types.py#L71
- Nheko (missing UI, using a slightly different format): https://github.com/Nheko-Reborn/mtxclient/commit/d6a0a4ebee83275dbbeb999679a22a7d238326ff and https://github.com/Nheko-Reborn/nheko/commit/978174af774f99bb70df2ad5307ae161be6190ff

Author: Partial implementation as a synapse module: https://github.com/matrix-org/msc4155-synapse-invite-filter

Reviewer: With my SCT hat:

The following implementations demonstrate the desire for filtering, and experiment with different ordering, but all demonstrate that the code is more than possible:

* https://github.com/element-hq/synapse/pull/18288
* https://github.com/Nheko-Reborn/mtxclient/commit/d6a0a4ebee83275dbbeb999679a22a7d238326ff
* https://github.com/Nheko-Reborn/nheko/commit/978174af774f99bb70df2ad5307ae161be6190ff
* https://github.com/famedly/synapse-invite-checker/blob/5608f40ffeaef1320edef8554cb432fa2c881fdd/synapse_invite_checker/types.py#L71
* https://github.com/matrix-org/msc4155-synapse-invite-filter

The following implementations show that there's desire, but don't do much with the actual config:
* https://github.com/element-hq/element-web/pull/29603

Collectively, I feel this satisfies the implementation component, though we should monitor for user feedback if the rule processing order feels wrong. If so, we can correct it with a future MSC relatively easily (it'll suck for people on implementations which use the 'old' rules, but as those implementations update users will see the changes).

Reviewer 4: >  If so, we can correct it with a future MSC relatively easily (it'll suck for people on implementations which use the 'old' rules, but as those implementations update users will see the changes).

I wish I shared your confidence. Won't it (also) suck for people who have set up rules with the "old" system, which suddenly don't work because the ordering has changed? And then they'll change the rules and get something that works with their client but not their server.

Reviewer 4: Yeah I don't disagree with you there. At some point we're going to have to just decide that we have enough confidence in the ordering that it's worth shipping. I just think it's worth having our eyes open to the fact that changing it later is going to be a bit of a buttpain, rather than "relatively easy" as Travis claimed.

Author: This proposal has been public and in discussion for a reasonable amount of time by now. I cannot currently think of further use cases that would lead to new rules in the configuration. I'm sure some exist but my hunch is that we won't discover them without the current proposal reaching wider usage. Therefore, my suggestion would be to commence this MSC through FCP and, if successful, into the spec and then evolve it from there. If we later decide that we'd better change things, we can version the event. This will cause some hassle for implementations and users but that feels better to me than getting stuck at this stage out of concern of future changes.

Reviewer 4: I remain somewhat concerned that nobody has implemented a useful UI for all these settings. Element Web has a very coarse "Allow invites: Y/N" which toggles whether `*` appears in the `blocked_users` list, but that's a far cry from a full UI.

As an example of the sort of things I'm worries about: what happens if one client decides to implement "block invites" by setting `blocked_servers: ["*"]`, while EW continues to do it via `blocked_users`? The two will be out of sync and give a very poor experience.

Honestly, if we're not expecting clients to implement a full UI, then I think we should scale back the specced proposal so that it only implements the bits that we *do* expect clients to implement.

Reviewer: I don't remember where it was documented, but the idea for an on/off toggle is to have a dedicated boolean in the object for it rather than manipulating specific settings. The idea behind this is a user may temporarily want to disable all invites without having to write down all of their other settings.

Reviewer 4: Thanks @Johennes, that's extremely helpful. I'm concerned, however, that it only (appears to) offer controls for `blocked_users` and `blocked_servers` (i.e, it does not expose `allowed_users`, `ignored_users`, `allowed_servers` or `ignored_servers`), and nor does it expose globbing.

It is the interplay between these settings that I think makes this proposal complicated to implement in a client, and my concern remains that we don't yet have a viable client implementation.

Author: No worries at all. 👍

specific users but doesn't have a mechanism to ignore entire servers. These shortcomings make the module
insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users [module] to allow filtering invites specifically. The scheme
outlined below is conceptually borrowed from the [gematik specification]. The main purpose of this proposal
is to ensure that this option is available for comparison with [other existing MSCs] that attempt to address
invite filtering.

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    "default": "allow | block",
    "user_exceptions": {
      "@someone:example.org": {},
      ...
    },
    "server_exceptions": {
      "example.org": {},
      ...
    }
  }
}
```

The `default` field defines the standard setting that is applied to all invites that don't match an
exception. Exceptions invert the default setting and are provided via the `user_exceptions` and

> `server_exceptions` fields which follow the format of the `ignored_users` field in `m.ignored_user_list`.

Reviewer 2: I think this comes with a problem of data loss. If you flip the "default" here, you end up with one of two problems. Either:

1. The exemptions are now blocked/allowed, so if I start with my invites allowed but block `@badguy:scam.org` and then flip the default then the only person who can talk to me is bad guy!
2. If the client flips the `default` and removes the exceptions then that data is lost forever and I need to reconstruct my list.

Could we make an easy change here where the exceptions are explicit about their behavior by including the block/allow for each key. While this looks redundant, it means flipping bits won't cause unexpected behavior here.

Author: Oh, I see! And sorry, I wasn't aware you're actually part of T&S. If you want to go ahead with putting the type into the exceptions, that sounds reasonable to me, like I said. I'll happily add this into the proposal if it turns out worthwhile in the implementation.

Also, since we have this in the gematik specification (though with a different event type), there must already be implementations of this somewhere, just apparently not in open source clients, sadly. 😕

Reviewer 2: The downside to the field is we need to specify what happens if a user appears twice? Do we have a priority ordering here?

Reviewer: I ended up implementing it as the following, with glob support:

1. Check `allow_` rules first - if allowed, stop processing rules and allow.
2. Check `ignore_` rules next - if ignored, stop processing rules and ignore.
3. Check `block_` rules next - if blocked, stop processing rules and block.
4. Finally, allow by default.

This enables a number of use cases (which can also apply to users):

1. "No invites at all" -> `blocked_servers: ["*"]`
2. "No invites except for example.com users" -> `blocked_servers: ["*"], allowed_servers: ["example.com"]`
3. "No invites from example.com" -> `blocked_servers: ["example.com"]`
4. "No invites from example.com unless it's Bob" -> `blocked_servers: ["example.com"], allowed_users: ["@bob:example.com"]`
5. "Allow all invites" -> `unset`
6. "No invites except for example.com, and ignore evil.example.com instead" -> `blocked_servers: ["*"], ignored_servers: ["evil.example.com"], allowed_servers: ["example.com"]`

UX-wise I'd expect that this would be simplified to a couple toggles and list instead:

* Allow invites? [on/off] -> `blocked_servers` and `allowed_servers` toggle with value `["*"]`
* Block or ignore invites? -> switch between `blocked_servers` and `ignored_servers`
* Exceptions: [user ID/server name lists] -> populate the inverse of the above two settings.

Users who want to get complex with it can edit their account data manually.

Reviewer: (if it makes more sense to write this up as a dedicated MSC, let me know)

Author: No, I'm happy to switch things around. The current proposal has the issue @Half-Shot pointed out, so that won't work anyway. Having `allowed_`, `blocked_`, ... fields seems slightly more flexible to me than our previous idea of keeping the top-level default and overriding it in each exception. This also feels similar to how server ACLs work which is nice.

I do wonder if we actually need to differentiate ignore and block here. Are there use cases where you don't want to be invited by somebody but you do care that they are informed about your rejection?

Reviewer: The primary use case is sometimes people see that you're blocking them and then they try to work around it so they can send more targeted abuse towards you. Ignoring swallows that initial signal, so attackers are (typically) less inclined to try targeting you.

Author: Sorry, I meant the other case. Is there a use case for letting somebody know that you're blocking them? If not then maybe we're fine with just `allowed_` and `ignored_` lists.

Author: Additionally, if we allow globs, do we still need to differentiate between users and servers? I think this

```json5
"allowed_servers": {
  "example.org": {}
}
```

should be equivalent to this

```json5
"allowed_users": {
  "*:example.org": {}
}
```

?

Author: @turt2live / @Half-Shot do you have any thoughts on whether or not to include a blocked list in addition to an ignored list and whether we'd be ok with a single list for both users and servers?

Reviewer: > Sorry, I meant the other case. Is there a use case for letting somebody know that you're blocking them? If not then maybe we're fine with just `allowed_` and `ignored_` lists.

Yes - users should have agency to decide whether they want to tell the other party whether the invite is explicitly rejected/not allowed. In future this may include something like "You're inviting the wrong account. Try `@me:example.org` instead".

> and whether we'd be ok with a single list for both users and servers?

I'd like to keep it separate for now, for ease of use by developers. Having too many 'tricks' like this makes the protocol feel more complicated than it is, sometimes.

Author: Thanks for expanding. I have updated proposal text accordingly.

Reviewer 5: > I'd like to keep it separate for now, for ease of use by developers. Having too many 'tricks' like this makes the protocol feel more complicated than it is, sometimes.

So basically it's because "you like it this way"? I can accept that but there should be some consesus  instead of your sole opinion, ain't it?

Seems like a clear duplicate functionality, and I guess the code would look like matching the sender of the invite with the list entries, so in this case _two_ lists instead of one, using the same matching. I don't see how this would make developers' life simpler.

There were mentions elsewhere about ordering (name matches first, server later, to make it possible to disable everone but allow someone), which is not mentioned here as a pro, but feels like it would be a mess anyway people guessing the double ordering (first names, then servers).

Using common globs (servers and users) does not feel complex to me, and clinet developers can create separate UI for servers if they feel like, I'm sure it's not very complex to understand/remember that `*:servername` is a "server" entry.

Reviewer: (not considering this thread as a blocker against the checklist in this case)

Reviewer 4: > So basically it's because "you like it this way"? I can accept that but there should be some consesus instead of your sole opinion, ain't it?

At the end of the day, this isn't a democracy; it's for the spec core team to consider the alternatives, taking community feedback into account, and reach a conclusion.

What *is* helpful, particularly in cases like this where there has been strong feedback in favour of a different implementation, is if the "Alternatives" section of the MSC text can set out those alternatives, together with the reason that the author of the MSC chose the approach they did. This helps avoid everyone asking "why not X", and helps the SCT and the rest of the community evaluate whether they agree with the author's assessment. @Johennes it might be useful to add some discussion of this issue to the Alternatives?

---

For my part, I also have a nagging feeling of the current datastructure having too many degrees of freedom, which could end up confusing users. It feels like we're only really supporting globs so that we can have a catch-all in `blocked_servers`, and maybe there are better ways to handle that. (e.g., reintroduce the `default: "block"|"allow"` key, to apply after everything else; or explicitly support `"*"` to mean "everything" rather than full glob functionality.)

I don't have a super strong opinion here, though. I think really, it's a case where we need something that can map onto a UI that makes sense for users, and as such I'm mostly interested from people who have implemented a UI for this, *particularly* if they have talked to a designer rather than taking the traditional Matrix approach of presenting the user with a flight console.

Author: > @Johennes it might be useful to add some discussion of this issue to the Alternatives?

Good call. Have extended the alternatives with https://github.com/matrix-org/matrix-spec-proposals/pull/4155/commits/be64f07349ee100d3153bfa502cbd938da0be79f.

As an example, a block-list configuration that ignores invites from `@badguy:scam.org` but allows invites
from any other user would look like this:

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    "default": "allow",
    "user_exceptions": {
      "@badguy:scam.org": {},
    }
  }
}
```

In contrast, an allow-list configuration that permits invites from users on goodguys.org but blocks invites
from all other servers would look like this:

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    "default": "block",
    "server_exceptions": {
      "goodguys.org": {},
    }
  }
}
```

Note that since the default setting for entities that don't match an exception is part of the configuration,
an exception for a user does _not_ need to be accompanied by an exception for their server[^1].

Servers MAY suppress invites for which the configuration evaluates to `block` and not send them to the recipient.
They MAY additionally reject the invite.

Clients SHOULD hide invites from users for which the permission configuration evaluates to `block`. They MAY

> allow reviewing ignored invites in a dedicated section of their UI.

Reviewer 6: It looks like both implementations (synapse PR and module) will refuse invite requests rather than suppressing or rejecting them. Invites must be signed by the recipient server, so the server has the ability to block the invite before it exists at all

In general, I think it makes more sense for servers to be the primary blocker and client-side hiding being an optional extra, i.e. servers SHOULD block invites, clients MAY hide invites (if the server doesn't have blocking implemented)

Reviewer: Some of this may also be clarified with https://github.com/matrix-org/matrix-spec-proposals/pull/4283

Author: I have added this into the proposal text.

## Potential issues

Larger permission configurations could run into the [event size limit] of 65536 bytes. This issue also exists

> with the `m.ignored_user_list` event.

Author: This is untrue actually. The event size limits only apply to room and state events but not account data.

Author: This has been removed.

# MSC4155: Invite filtering

Matrix supports ignoring users via the eponymous [module] and the `m.ignored_user_list` account data
event. This is a nuclear option though and will suppress both invites and room events from the ignored
users. Additionally, the `m.ignored_user_list` event only allows for block-list configurations that ignore
specific users but doesn't have a mechanism to ignore entire servers. These shortcomings make the module
insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users [module] to allow filtering invites specifically. The scheme
outlined below was conceptually borrowed from the [gematik specification].

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    // User-level settings
    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],

>     // Server-level settings

Reviewer 5: What is the point of having separate "server level settings" when the user level settings are globs?
`{ "blocked_users": "*:badguys.org" }` is the same as `{ "blocked_servers": "badguys.org" }`, isn't it?

Reviewer: see https://github.com/matrix-org/matrix-spec-proposals/pull/4155#discussion_r2030686327

    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],

>     "blocked_servers": [ "*" ] // Glob expressions allowed

Reviewer 2: The text below says yes, I think the examples could be clearer

>     "blocked_servers": [ "*" ] // Glob expressions allowed

Author: Yes, you're right. I have accepted Travis' suggestion which should fix this.

  }
}
```

All properties in `content` are optional arrays. A missing or `null` property MUST be treated like an
empty array. The array elements are [glob expressions] to be matched against user IDs and server names,
respectively.

> When evaluating an invite against the contents of `m.invite_permission_config`, implementations should

Reviewer: ```suggestion
When evaluating an invite against the contents of `m.invite_permission_config`, implementations MUST
```

apply the following order:

1.  If the invite matches any `allow_*` setting, stop processing and allow.
2.  If the invite matches any `ignore_*` setting, stop processing and ignore.

> 3.  If the invite matches any `block_*` setting, stop processing and block.

Reviewer: I agree with checking users before servers, for those reasons exactly. 

The logic would become:
1. For each `_users` setting:
  1. If it matches `allow_users`, stop processing and allow.
  2. If it matches `ignore_users`, stop processing and ignore.
  3. If it matches `block_users`, stop processing and block.
2. For each `_servers` setting:
  1. Repeat 1.1-1.3, but using `*_servers`
3. Otherwise, allow.

> 3.  If the invite matches any `block_*` setting, stop processing and block.

Reviewer 2: This is what I implemented because I was so certain it felt right I didn't check the actual MSC. @Johennes do you concur, or have thoughts?

> 3.  If the invite matches any `block_*` setting, stop processing and block.

Reviewer 5: Since you choose not to do a longest pattern match what is the order of the evaluation of the various glob patterns in the same category, like `allow_users`? Order of definition, first match? 

So the clients are expected to possess the feature to reorder the lists?

> 3.  If the invite matches any `block_*` setting, stop processing and block.

Reviewer: They are JSON arrays, so order is maintained through deserialization. The order is as presented in that array.

> 3.  If the invite matches any `block_*` setting, stop processing and block.

Author: Sounds good, I hadn't thought of that case. Addressed with https://github.com/matrix-org/matrix-spec-proposals/pull/4155/files#r2115421415.

> 1.  If the invite matches any `allow_*` setting, stop processing and allow.
> 2.  If the invite matches any `ignore_*` setting, stop processing and ignore.
> 3.  If the invite matches any `block_*` setting, stop processing and block.

Author: ```suggestion
1.  Verify the invite against each `*_users` setting: 
  1.  If it matches `allowed_users`, stop processing and allow.
  2.  If it matches `ignored_users`, stop processing and ignore.
  3.  If it matches `blocked_users`, stop processing and block.
2.  Verify the invite against each `*_servers` setting: 
  1.  If it matches `allowed_servers`, stop processing and allow.
  2.  If it matches `ignored_servers`, stop processing and ignore.
  3.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.
```

feedback to the inviter whereas blocking rejects the invite.

The semantics and order of evaluation enable a number of use cases, for instance:

```json5
// Invites from everyone
{
  "type": "m.invite_permission_config",
  "content": { }
}

// No invites at all
{
  "type": "m.invite_permission_config",
  "content": {
    "blocked_servers": [ "*" ]
  }
}

// Only invites from goodguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "allowed_servers": [ "goodguys.org" ],
    "blocked_servers": [ "*" ]
  }
}

// Invites from everyone except badguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "blocked_servers": [ "badguys.org" ]
  }
}

// No invites from badguys.org unless it's @goodguy:badguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "allowed_users": [ "@goodguy:badguys.org" ],
    "blocked_servers": [ "badguys.org" ]
  }
}

// Only invites from goodguys.org and don't provide feedback to reallybadguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "allowed_servers": [ "goodguys.org" ],
    "ignored_servers": [ "reallybadguys.org" ],
    "blocked_servers": [ "*" ]
  }
}
```

> 

Reviewer 5: (The examples blissfully ignore globbing, which would make them possibly painfully complex.)

Reviewer: This can be covered at spec writing time and isn't essential for the MSC, in my opinion.

The author can choose to enhance the examples if they wish.

Author: I'm happy to add more examples if it helps but thought the current ones would already be quite comprehensive. If you have concrete ideas, feel free to comment them as suggestions on the PR directly.

Servers SHOULD enforce `m.invite_permission_config` against incoming invites but MAY also delegate
the task to clients to allow them to let users review ignored and blocked invites. Clients MUST
enforce the permission config, if their server doesn't do it. In either case, servers MUST communicate

> their behaviour to clients through a new capability `m.invite_permission_config_enforced`.

Reviewer 5: This is not good. 

By default the server MUST enforce permissions against **all open invites** also, so, for example, blocking a user should reject **all** invites from the specific user. Otherwise this feature cannot ensure removal of **mass invites** from spammers (which otherwise would be impossible due to rate limiting and the room-oriented way of the current API).

If enforcement is not preferred there could be a way to flag that to the server, though I don't readily see a situation where the user wanted to keep invites from ignored users.

The same is probably true for "incoming invites" (I guess that means future incoming invites?) as well.

Reviewer: This MSC doesn't prohibit a client or server from taking action against existing invites. Servers can choose to take action when the user sets a config, or leave it up to the client (the common case). For spammy invites, servers already can and do remove the invites - this MSC doesn't change that.

Reviewer 5: @Johennes in your opinion is there any point to accept this MSC allowing server behaviour deviation and have a later MSC invalidate this? Do you think this is out of scope for this MSC?

Reviewer 5: > but if it doesn't, then the client MUST do it instead.

And the client cannot, since it will be rate-limited and the client basically would kick itself off the server.
Basically the whole point of https://github.com/matrix-org/matrix-spec-proposals/pull/4294 was this sentence:

> Servers are **required** to reject all pending invites from all the ignored inviter users.

For me it's weird to create _another_ endpoint when this one could perfectly (and logically) handle that.

But if I understand correctly the suggested way is to create an MSC against this MSC containing that one sentence. That feels extremely weird to me as well. But of course I can do that…

Reviewer 7: (I haven't read every last word in this thread.) Overall I agree this proposal is very weird to not enforce this as a server. One of the mantras of Matrix is complex servers, simple clients. Servers are the only entity here with enough information to process these efficiently, let's just enforce that it MUST be done.

What benefit is there in pushing clients to do it? I only see downsides.

Reviewer 7: Isn't users reviewing it essentially what we have today though with the invite system?

Reviewer 8: > I think the only benefit of the client doing it is that the user could review ignored invites. If the server does it, the user wouldn't notice anything. Maybe this is actually desirable but some of the other invite filtering MSCs were designed with client-side logic which is why I thought there was a need for this.

IMHO this is really important. A user decides to block every incoming invite. But then he gets the information over another communication channel, that he has been invited to an important room. How should he be able to review the rejected invites and join instead, when the server already enforces the invite filtering.

Reviewer 2: > IMHO this is really important. A user decides to block every incoming invite. But then he gets the information over another communication channel, that he has been invited to an important room. How should he be able to review the rejected invites and join instead, when the server already enforces the invite filtering?

This has happened to me already :). They pinged me to say they were unable to invite me to a room and I allow-listed their mxid so they could invite me.

This wouldn't be the case for an ignore, but ignore is a fairly powerful tool for bad actors and not something you'd (typically) have a blanket rule for. I think in these cases if something is important then the inviter can always reach out to the user via other means.

Author: I have added @benkuly's concern in the potential issues section. I'll leave this thread resolved. Any further discussion can still happen on the potential issues section.

Reviewer 5: If I understand correctly when a user **blocks** invites from a source (using this API) then they **cannot** review **future** incoming the invites, no matter whether the reserver removes past invites or not: server blocks all invites "at the source", they do not reach neither the room nor the user. 
(As it was mentioned the whole point of this API is to block/reject invites without any user review or disturbance. If the user want to review the invites then they simply do not use this API at all. Then they will have the problem of not being able to mass remove them, but that's out of both my and MSCs scope. :-))

When a user **ignores** the invites then, if I understand correctly, they are sent to the room, so they will be visible for the user if they check hidden room events, but hese are not sent through the sync to the user and generate no notifications. Ignore does not (and would not) affect pending invites, if I understand the intent correctly. (This way the user does not get notifications but still can review invites by browsing room history, right?)

```json5
{
  "capabilities": {
    "m.invite_permission_config_enforced": {
      "enabled": true
    }
  }
}
```

> If `enabled` is set to `true`, the server enforces the permission config. If the capability is missing,

Reviewer 5: What is the base of decision for the value of this flag? Server code? Server configation?

Reviewer: It's an implementation detail.

Reviewer: If we find that the level of usage does not match our expectations, we can put forward an MSC which removes the capability, thus making the behaviour required. For now, we prefer to have optionality.

clients SHOULD behave as if `enabled` was set to `false`.

## Potential issues

None.

> 

Reviewer 5: - Is there are limit of the entries? 
- Can the **same** entry included in multiple lists? Should it?
- What does the `reason` contain for the blocked invites? 
- How should servers handle when `ignored_users` and these allow/ignore/blocks overlap? Don't care?

Reviewer 2: Reasons sound like an implementation matter imo. For serverside there aren't any reasons because: 

- When you block an invite, it's not a rejection. It's the sever sending a 403 to your request for an invite so the actual invite is never sent.
- An ignored invite obviously is just silently accepted and ignored, so there is no response.

The clientside sounds like an implementation detail, since divergence isn't a problem here. It's okay for some clients to give a reason and some not to, because no software should be using the reason field to drive anything automatically. Reasons are solely for humans to consume, so they should remain freeform.

---

With regards to the overlap, Travis was saying there is a stack rank order of operations of:

- accept userignore user, block user, accept server, ignore server, block server. In each of these categories the implementation should iterate through the list and as soon as it gets a match apply the appropriate rule. This is deterministic, as we have determined that each of these

Reviewer 9: Isn't the overlap question actually about how this MSC interacts with the `m.ignored_user_list` account data? Does `m.ignored_user_list` have priority over this setting in regards to ignoring invites, to keep the current behavior?

Author: Hm, either that or maybe we take invites out of `m.ignored_user_list` to eliminate the overlap? 🤔

    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],
    // Server-level settings
    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],
    "blocked_servers": [ "*" ] // A feature of all the fields at this level, globs
  }
}
```

All properties in `content` are optional arrays. A missing or `null` property MUST be treated like an
empty array. The array elements are [glob expressions] to be matched against user IDs and server names,
respectively.

When evaluating an invite against the contents of `m.invite_permission_config`, implementations MUST
apply the following order:

1.  Verify the invite against each `*_users` setting: 
  1.  If it matches `allowed_users`, stop processing and allow.
  2.  If it matches `ignored_users`, stop processing and ignore.
  3.  If it matches `blocked_users`, stop processing and block.
2.  Verify the invite against each `*_servers` setting: 
  1.  If it matches `allowed_servers`, stop processing and allow.
  2.  If it matches `ignored_servers`, stop processing and ignore.
  3.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.

> 4.  Otherwise allow.

Reviewer: (it's important to `otherwise, allow` in life 😇)

The semantics of "ignore" and "block" follow [MSC4283] which means ignoring hides the invite with no

> feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Reviewer 2: ```suggestion
feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Servers which apply invite filter rules MUST block invites by responding to the following endpoints with an error:

- `PUT /_matrix/federation/(v1|v2)/invite/{roomId}/{eventId}`
- `POST /_matrix/client/v3/rooms/{roomId}/invite`
- `POST /_matrix/client/v3/createRoom` (checking the `invite` list)
- `PUT /_matrix/client/v3/rooms/{roomId}/state/m.room.member/{stateKey}` (for invite membership)

The error SHOULD be `M_INVITE_BLOCKED` with a HTTP 403 status code.

For ignoring invites, these endpoints MUST return identical response bodies and approximately identical response times to a true successful invite.
```

> feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Reviewer 2: ^ Area for concern is we should spell out exactly how ignore should work then :)

> feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Reviewer: @Johennes this is currently blocking the remaining two concerns from becoming resolved :)

> feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Reviewer 2: Suggesting https://github.com/matrix-org/matrix-spec-proposals/pull/4155#discussion_r2120384340 as a decent attempt to cover this.

> feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

Author: Looks great, thank you!

apply the following order:

1.  Verify the invite against each `*_users` setting: 
  1.  If it matches `allowed_users`, stop processing and allow.
  2.  If it matches `ignored_users`, stop processing and ignore.
  3.  If it matches `blocked_users`, stop processing and block.
2.  Verify the invite against each `*_servers` setting: 
  1.  If it matches `allowed_servers`, stop processing and allow.
  2.  If it matches `ignored_servers`, stop processing and ignore.
  3.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.

The semantics of "ignore" and "block" follow [MSC4283] which means ignoring hides the invite with no
feedback to the invite whereas blocking rejects (or refuses, in the case of servers) the invite.

When blocking an inviter, the server must respond to the following endpoints with an error:

- `PUT /_matrix/federation/(v1|v2)/invite/{roomId}/{eventId}`
- `POST /_matrix/client/v3/rooms/{roomId}/invite`
- `POST /_matrix/client/v3/createRoom` (checking the `invite` list)
- `PUT /_matrix/client/v3/rooms/{roomId}/state/m.room.member/{stateKey}` (for invite membership)

The error SHOULD be `M_INVITE_BLOCKED` with a HTTP 403 status code.

> When ignoring an invite, these endpoints MUST handle an invite normally as if accepted. However, the server MUST not include the invite down client synchronization endpoints such as  `GET /_matrix/client/v3/sync` or MSC4186's sliding sync endpoint. In addition, these invites MUST be ignored when calculating push notifications.

Author: (Just line-wrapping for easier review by others.)

```suggestion
When ignoring an invite, these endpoints MUST handle an invite normally as if accepted. However, the server
MUST not include the invite down client synchronization endpoints such as `GET /_matrix/client/v3/sync` or
MSC4186's sliding sync endpoint. In addition, these invites MUST be ignored when calculating push notifications.
```

  3.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.

The semantics of "ignore" and "block" follow [MSC4283] which means ignoring hides the invite with no
feedback to the invite whereas blocking rejects (or refuses, in the case of servers) the invite.

When blocking an inviter, the server must respond to the following endpoints with an error:

- `PUT /_matrix/federation/(v1|v2)/invite/{roomId}/{eventId}`
- `POST /_matrix/client/v3/rooms/{roomId}/invite`
- `POST /_matrix/client/v3/createRoom` (checking the `invite` list)
- `PUT /_matrix/client/v3/rooms/{roomId}/state/m.room.member/{stateKey}` (for invite membership)

The error SHOULD be `M_INVITE_BLOCKED` with a HTTP 403 status code.

When ignoring an invite, these endpoints MUST handle an invite normally as if accepted. However, the server
MUST not include the invite down client synchronization endpoints such as `GET /_matrix/client/v3/sync` or
MSC4186's sliding sync endpoint. In addition, these invites MUST be ignored when calculating push notifications.

Otherwise, all other endpoints (such as `GET /_matrix/client/v3/rooms/{roomId}/state`) should work as before. 

> Servers are not expected to process these rules for appservice users when calculating events to send down `PUT /_matrix/app/v1/transactions`. Appservices are not expected to be run directly by client users, and should be able to handle their own spam prevention.

Author: (Just line-wrapping for easier review by others.)

```suggestion
Servers are not expected to process these rules for appservice users when calculating events to send down
`PUT /_matrix/app/v1/transactions`. Appservices are not expected to be run directly by client users, and
should be able to handle their own spam prevention.
```

users. Additionally, the `m.ignored_user_list` event only allows for block-list configurations that ignore
specific users but doesn't have a mechanism to ignore entire servers. These shortcomings make the module
insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users [module] to allow filtering invites specifically. The scheme
outlined below was conceptually borrowed from the [gematik specification].

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    // User-level settings
    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],
    // Server-level settings
    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],
    "blocked_servers": [ "*" ] // A feature of all the fields at this level, globs
  }
}
```

All properties in `content` are optional arrays. A missing or `null` property MUST be treated like an

> empty array. The array elements are [glob expressions] to be matched against user IDs and server names,

Reviewer 4: The potential confusion is about whether the user id is split into local and server parts before applying wildcards. Doing so might help people avoid making matches more general than they meant to. But, I'm not arguing for that, just asking that we be explicit that the globbing treats the whole ID as a single string.

> You probably actually would want `*.matrix.org` if you're trying to get `dendrite.matrix.org`?

You might well, but the question isn't "how do I write a glob if I want to match a given set of IDs", but rather the opposite direction: "given a glob, possibly written by a user who doesn't quite know what they are doing, are we clear about what exactly it will match?"

respectively.

When evaluating an invite against the contents of `m.invite_permission_config`, implementations MUST
apply the following order:

1.  Verify the invite against each `*_users` setting: 
  1.  If it matches `allowed_users`, stop processing and allow.
  2.  If it matches `ignored_users`, stop processing and ignore.
  3.  If it matches `blocked_users`, stop processing and block.
2.  Verify the invite against each `*_servers` setting: 
  1.  If it matches `allowed_servers`, stop processing and allow.
  2.  If it matches `ignored_servers`, stop processing and ignore.
  3.  If it matches `blocked_servers`, stop processing and block.

> 3.  Otherwise, allow.

Reviewer: it feels beneficial to follow the [Server ACL](https://spec.matrix.org/v1.15/client-server-api/#server-access-control-lists-acls-for-rooms) order, with a slight change to the very last rule. Instead of "Otherwise, deny", we "Otherwise, allow" for backwards compatibility. 

I think this means processing the rules as such:

1. Verify the invite against each `*_users` setting:
   1. If it matches `blocked_users`, stop processing and block.
   2. If it matches `ignored_users`, stop processing and ignore.
   3. If it matches `allowed_users`, stop processing and allow.
2. Verify the invite against each `*_servers` setting:
   1. If it matches `blocked_servers`, stop processing and block.
   2. If it matches `ignored_servers`, stop processing and ignore.
   3. If it matches `allowed_servers`, stop processing and allow.
3. Otherwise, allow.

I don't think we want to check each in batches (all blocks, then all ignores, then all allows) specifically to ensure the user is able to allow-list certain senders from otherwise banned servers. For example: `blocked_servers: [matrix.org], allowed_users: [@abuse:matrix.org]`. 

Later, when we have more rules, we'd append to just before the "Otherwise, allow" rule.

> 3.  Otherwise, allow.

Reviewer 10: Consider the following issue[^1]. 

1. `@alice:example.com` is added to `allowed_users`.
2. `example.com` becomes hostile.
3. `example.com` is added to `blocked_servers`.
4. `@alice:example.com` sends malicious invitations.

When `example.com` is added to `blocked_servers`, the entry for each user from that server in `allowed_users` becomes stale. There could probably be other scenarios where entries can become stale too.

A way to fix this could be to instead  provide a `*_users` object for each`server`.

```
{
  "example.com": { "allowed_users": "@alice:example.com" }
}
```

You would then also need to be able to state whether all users from the server can send invitations, or specific users. So you would also need to introduce a "match all" rule for `*_users`. 

But it's up to you how much of a problem you see this to be, and again the issue is specific to trying to make the proposal closer to server ACL. 

[^1]: You can't avoid this problem by inverting the behavior described in the comment above (ie requiring that both checks for servers and users pass)  because then you would end up in a situation where the `allowed_users` rule can become stale as the `allowed_servers` rules are changed.

present proposal is functionally inferior to some of the alternatives outline there. It does, for
instance, not cover the change history of the permission config or sharing the config among different
users. The proposal is, however, a straightforward and easy to implement extension of the existing
`m.ignored_user_list` mechanism. See also [this comment] for further details. This proposal is additionally
extensible for further types of blocking in the future. For example, a future MSC could create definitions
and behaviours to block/ignore/allow invites from contacts, of a particular type (DM, space, etc), 
to a particular room, or even with given keywords.

## Security considerations

None.

## Unstable prefix

Until this proposal is accepted into the spec, implementations should refer to `m.invite_permission_config`
and `m.invite_permission_config_enforced` as `org.matrix.msc4155.invite_permission_config` and
`org.matrix.msc4155.invite_permission_config_enforced`, respectively. Note that the [gematik specification],
which predates this MSC, uses an event type of `de.gematik.tim.account.permissionconfig.v1` and
a different event schema.

> 

Reviewer 2: ```suggestion
The error `M_INVITE_BLOCKED` should be `ORG.MATRIX.MSC4155.M_INVITE_BLOCKED` until this proposal is accepted into the spec.
```

insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users [module] to allow filtering invites specifically. The scheme
outlined below was conceptually borrowed from the [gematik specification].

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    // User-level settings
    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],
    // Server-level settings
    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],
    "blocked_servers": [ "*" ] // A feature of all the fields at this level, globs
  }
}
```

All properties in `content` are optional arrays. A missing or `null` property MUST be treated like an
empty array. The array elements are [glob expressions] to be matched against user IDs and server names,
respectively.

When evaluating an invite against the contents of `m.invite_permission_config`, implementations MUST

> apply the following order:

Reviewer 7: Is this evaluated after `m.ignored_user_list`?

Author: Good question, we should definitely clarify the order.

I don't have a strong opinion but `m.ignored_user_list` seems like the bigger hammer given that it will also suppress events from ignored users in existing rooms. So I'd lean towards evaluating that before `m.invite_permission_config`.

WDYT?

Reviewer 7: I think that makes sense, yep. If you've completed ignored a user it would seem weird to potentially *allow* an invite from them by evaluating this first.

Also, I think that's more backwards-compatible.

Author: Have encoded this into the proposal text.

1.  Verify the invite against each `*_users` setting:
    1.  If it matches `allowed_users`, stop processing and allow.
    2.  If it matches `ignored_users`, stop processing and ignore.
    3.  If it matches `blocked_users`, stop processing and block.
2.  Verify the invite against each `*_servers` setting:
    1.  If it matches `allowed_servers`, stop processing and allow.
    2.  If it matches `ignored_servers`, stop processing and ignore.
    3.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.

The semantics of "ignore" and "block" follow [MSC4283] which means ignoring hides the invite with no

> feedback to the invite whereas blocking rejects (or refuses, in the case of servers) the invite.

Reviewer 7: ```suggestion
feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.
```

When blocking an inviter, the server must respond to the following endpoints with an error:

- `PUT /_matrix/federation/(v1|v2)/invite/{roomId}/{eventId}`
- `POST /_matrix/client/v3/rooms/{roomId}/invite`
- `POST /_matrix/client/v3/createRoom` (checking the `invite` list)
- `PUT /_matrix/client/v3/rooms/{roomId}/state/m.room.member/{stateKey}` (for invite membership)

The error SHOULD be `M_INVITE_BLOCKED` with a HTTP 403 status code.

When ignoring an invite, these endpoints MUST handle an invite normally as if accepted. However, the server
MUST not include the invite down client synchronization endpoints such as `GET /_matrix/client/v3/sync` or

> MSC4186's sliding sync endpoint. In addition, these invites MUST be ignored when calculating push notifications.

Reviewer 7: Is the expectation that the state of an ignored user just stays at invited?

Are there edge cases to consider:

If a user unblocks the sender in the future? Are they able to re-invite? Do they just start seeing the invite?

Reviewer 2: Ah yes, so currently it's only initial syncs for ignored users. I suppose it's a tricky thing implementation wise since you need to go and retroactively fetch invites from before your current sync position if the config changes, which is a bit bleh.

Reviewer 7: Probably not important, but probably needs to be defined.

# MSC4155: Invite filtering

> Matrix supports ignoring users via the eponymous [module] and the `m.ignored_user_list` account data

Reviewer 11: Thanks!

users. Additionally, the `m.ignored_user_list` event only allows for block-list configurations that ignore
specific users but doesn't have a mechanism to ignore entire servers. These shortcomings make the module
insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users [module] to allow filtering invites specifically. The scheme
outlined below was conceptually borrowed from the [gematik specification].

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    // User-level settings
    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],
    // Server-level settings
    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],
    "blocked_servers": [ "*" ] // A feature of all the fields at this level, globs
  }
}
```

All properties in `content` are optional arrays. A missing or `null` property MUST be treated like an

> empty array. The array elements are [glob expressions]. Any `*_users` glob is to be matched against

Reviewer 11: If we're explicitly extending to include `null` we should be complete, ie. `undefined` or indeed anything that isn't an array should be treated as absent (if we just said 'missing' I'd assume it was implicit, but if you're calling out one I think you need to be complete).

Reviewer 7: I'd prefer only missing to be considered empty. What's the goal of including null here?

Author: > If we're explicitly extending to include null we should be complete, ie. undefined or indeed anything that isn't an array should be treated as absent [...]

I think `undefined` is just the JS equivalent of "missing" in JSON but good point on other data types.

> What's the goal of including null here?

There's occasionally been complaints that the spec doesn't clearly describe the behavior for `null`, missing or empty values. So I was trying to be specific here with the intention that a `null` or missing value doesn't have any special meaning.

However, I suppose that's sort of obvious because how else would such a value be interpreted? I have removed the sentence entirely with 5c3cab864b0903687c22f492944c0ada5b309dcc.

Reviewer 7: I think the options are that invalid data should be treated as empty or that it invalidates the whole object. 🤷

Author: I wonder if this is something that could be defined spec-wide?

FWIW I'm dealing with this when implementing https://github.com/element-hq/synapse/pull/18195 as well.

Reviewer: It may be a spec-wide problem, but being explicit in the MSC in the meantime helps ensure it's covered and not delayed behind yet more MSCs.

Author: Cool. 👍 Have added it into the proposal with f17d2ee1c672524f0318d17eb457f6152524c1f0.

full user IDs (localpart and domain). Any `*_servers` glob is to be matched against server names /

> domain parts of user IDs.

Reviewer: +1 we should follow ACL semantics here and exclude port when matching

  "content": {
    "allowed_servers": [ "goodguys.org" ],
    "blocked_servers": [ "*" ]
  }
}

// Invites from everyone except badguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "blocked_servers": [ "badguys.org" ]
  }
}

// Only invites from goodguys.org except for @notactuallyguy:goodguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "blocked_users": [ "@notactuallyguy:goodguys.org" ],
    "allowed_servers": [ "goodguys.org" ]
    "blocked_servers": [ "*" ]
  }
}

// No invites from badguys.org unless it's @goodguy:badguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "allowed_users": [ "@goodguy:badguys.org" ],
    "blocked_servers": [ "badguys.org" ]
  }
}

// Only invites from goodguys.org and don't provide feedback to reallybadguys.org
{
  "type": "m.invite_permission_config",
  "content": {
    "allowed_servers": [ "goodguys.org" ],
    "ignored_servers": [ "reallybadguys.org" ],
    "blocked_servers": [ "*" ]
  }
}
```

Servers MUST enforce `m.invite_permission_config` against incoming new invites. Additionally, Servers
SHOULD apply the config against existing pending invites as well.

> ## Potential issues

Reviewer: Last point from the T&S meeting: from a UX perspective, where clients expose more complicated interfaces for managing these settings, they SHOULD also inform users of changes. For example, if the user already allowed `@alice:example.org` to send them invites and they go to add a block for `example.org` as a server, they should be told that "Alice can still send you invites despite this setting". The user can correct this by removing either the server-level block or the allowlist entry for Alice.

Reviewer 2: Yep, makes sense. This places a higher burden on clients to calculate the result of the values they set, but I think it's important to do so.

communication channel, they'll have to update their permission configuration and request a re-invite.

## Alternatives

Instead of introducing a separate account data event, the existing `m.ignored_user_list` could have
been expanded. This would, however, not only effect invites but also events in existing rooms which
makes it a much more nuclear option. Additionally, the existing schema of `m.ignored_user_list`
complicates morphing it into something that optionally supports allow-list semantics.

Regarding `m.invite_permission_config`, the split between user settings and server settings is
technically not needed because glob expressions are powerful enough to allow matching either.
Splitting them is more explicit and prevents unintended globbing mistakes, however. The fact that
a user glob and a server glob can overlap does not seem problematic because this proposal includes
a deterministic processing order for all settings.

A previous version of this proposal included a base setting of block / allow all with user and
server exceptions applied on top. In this scheme, flipping the base setting also inverts the semantics

> of all exceptions, however, which makes changing the configuration quite complicated.

Author: Sounds good. I have added `enabled` with 8c30b5639b005935f8da683904a3887a739f9752.

specific users but doesn't have a mechanism to ignore entire servers. These shortcomings make the module
insufficient for use cases such as tightly locked down applications where ignoring needs to be the default
behaviour.

This proposal generalises the ignoring users module to allow filtering invites specifically. The scheme
outlined below was conceptually borrowed from the [gematik specification].

## Proposal

To allow users to configure which other users are allowed to invite them into rooms, a new account data
event `m.invite_permission_config` is introduced.

```json5
{
  "type": "m.invite_permission_config",
  "content": {
    // Global on/off toggle
    "enabled": true,
    // User-level settings
    "allowed_users": [ "@john:goodguys.org", ... ],
    "ignored_users": [ ... ],
    "blocked_users": [ ... ],
    // Server-level settings
    "allowed_servers": [ ... ],
    "ignored_servers": [ ... ],
    "blocked_servers": [ "*" ] // A feature of all the fields at this level, globs
  }
}
```

`enabled` is a boolean property and defaults to `true` if omitted. It provides clients with a convenience on/off

> toggle that lets them deactivate the configuration without purging it.

Reviewer 4: @Half-Shot, @turt2live: as those arguing for `enabled: false` to be a high-level "block all invites": can you weigh in on this please?

All other properties in `content` are optional arrays. A missing property MUST be treated like an empty array.
Any value that is not an array of strings MUST lead to the entire event being considered invalid.

> The array elements are [glob expressions]. Any `*_users` glob is to be matched against full user IDs (localpart

Reviewer 12: There doesn't seem to be a way to match a user with a literal "*" or "?" in their user ID without also matching other users.  e.g. if I wanted to ignore a user `@*:example.com`, I could ignore `@*:example.com`, which would ignore everyone from `example.com`, or `@?:example.com`, which would ignore all single-character users from `example.com`.

Maybe we need backslash-escaping for the special glob characters?

(This isn't a problem with server names because of course `*` and `?` aren't valid in server names)

    2.  If it matches `allowed_users`, stop processing and allow.
    3.  If it matches `ignored_users`, stop processing and ignore.
    4.  If it matches `blocked_users`, stop processing and block.
    5.  If it matches `allowed_servers`, stop processing and allow.
    6.  If it matches `ignored_servers`, stop processing and ignore.
    7.  If it matches `blocked_servers`, stop processing and block.
3.  Otherwise, allow.

The semantics of "ignore" and "block" follow [MSC4283] which means ignoring hides the invite with no
feedback to the inviter whereas blocking rejects (or refuses, in the case of servers) the invite.

When blocking an inviter, the server MUST respond to the following endpoints with an error:

- `PUT /_matrix/federation/(v1|v2)/invite/{roomId}/{eventId}`
- `POST /_matrix/client/v3/rooms/{roomId}/invite`
- `POST /_matrix/client/v3/createRoom` (checking the `invite` list)
- `PUT /_matrix/client/v3/rooms/{roomId}/state/m.room.member/{stateKey}` (for invite membership)

The error SHOULD be `M_INVITE_BLOCKED` with a HTTP 403 status code.

When ignoring an invite, these endpoints MUST handle an invite normally as if accepted. However, the server

> MUST not include the invite down client synchronization endpoints such as `GET /_matrix/client/v3/sync` or

Reviewer 4: ```suggestion
MUST NOT include the invite down client synchronization endpoints such as `GET /_matrix/client/v3/sync` or
```

MSC4186's sliding sync endpoint. In addition, these invites MUST be ignored when calculating push notifications.
For clarity, this means that the invited user will get a regular `invite` membership event in the target room

> but will never be notified about that event. Similar, to `m.ignored_user_list`, clients SHOULD perform an

Reviewer 12: Nit: there seems to be a conflict here, in that one sentence says that the invited user will "never" be notified about the event, but the next sentence suggests that they might get it if they later change the `m.invite_permission_config`.

```suggestion
but will never be notified about that event, unless they modify their `m.invite_permission_config` such that the invite is no longer ignored. Similar, to `m.ignored_user_list`, clients SHOULD perform an
```

a deterministic processing order for all settings.

A previous version of this proposal included a base setting of block / allow all with user and
server exceptions applied on top. In this scheme, flipping the base setting also inverts the semantics
of all exceptions, however, which makes changing the configuration quite complicated.

A comprehensive comparison of existing invite filtering proposals may be found in [MSC4192]. The
present proposal is functionally inferior to some of the alternatives outline there. It does, for
instance, not cover the change history of the permission config or sharing the config among different
users. The proposal is, however, a straightforward and easy to implement extension of the existing
`m.ignored_user_list` mechanism. See also [this comment] for further details. This proposal is additionally
extensible for further types of blocking in the future. For example, a future MSC could create definitions
and behaviours to block/ignore/allow invites from contacts, of a particular type (DM, space, etc), 
to a particular room, or even with given keywords.

## Security considerations

> None.

Reviewer 13: Depending on the implementation strategy, the glob evaluation could be vulnerable to ReDoS or other complexity/performance risks.
